### ISSUE TITLE
qvm-pool/volume CLI tools and files report unrealistic space usage for disposable qubes

### ISSUE THREAD

**Issue by emanruse** [QubesOS/qubes-issues#8704]
### Qubes OS release

4.1.2

### Brief summary

`qvm-*` command line tools report unrealistic storage usage for disposable qubes.

### Steps to reproduce

Create a 1 GiB RAM based pool:

```
user@dom0:~ > mkdir -p ~/tmp/mydir
user@dom0:~ > sudo mount --types tmpfs --options size=1G my_ram_pool /home/user/tmp/mydir
user@dom0:~ > qvm-pool add my_ram_pool file --option revisions_to_keep=1 --option dir_path="${HOME}/tmp/mydir"
```

Create a named disposable qube:

```
user@dom0:~ > qvm-create --class DispVM AAA -P my_ram_pool --label=red
user@dom0:~ > qvm-pool info my_ram_pool | grep usage
usage               0
```

Start the qube:

```
user@dom0:~ > qvm-run AAA xterm
```

and create a 500 MiB in it:

```
user@AAA:~ > dd if=/dev/urandom of=~/500mb.img bs=4k iflag=fullblock,count_bytes count=500M
128000+0 records in
128000+0 records out
524288000 bytes (524 MB, 500 MiB) copied, 1.59195 s, 329 MB/s
user@AAA:~ > ls -l 500mb.img 
-rw------- 1 user user 524288000 2023-11-11 14:11:37 500mb.img
```

Check space usage in dom0:

```
user@dom0:~ qvm-pool info my_ram_pool | grep usage
usage               8192

user@dom0:~ > df -h ~/tmp/mydir/
Filesystem      Size  Used Avail Use% Mounted on
my_ram_pool     1.0G  8.0K  1.0G   1% /home/user/tmp/mydir

user@dom0:~ > ls -l ~/tmp/mydir/appvms/AAA/volatile.img 
-rw-r--r-- 1 root root 12884901888 2023-11-11 14:10:25 /home/user/tmp/mydir/appvms/AAA/volatile.img

user@dom0:~ > du -hs ~/tmp/mydir/appvms/AAA/volatile.img 
8.0K	/home/user/tmp/mydir/appvms/AAA/volatile.img

user@dom0:~ > qvm-volume info AAA:volatile | grep -E '(size|usage)'
size               12884901888
usage              8192
```

### Expected behavior

The numbers reported by the `qvm-` tools should match actuality. Then one can know how much space is left in a pool etc.

### Actual behavior

For any running DispVM the reported usage is always 8 KiB and the reported volatile volume size (as well as volatile.img file size) is always 12 GiB.

12 GiB is obviously much more than the actual capacity of the pool (only 1 GiB). The actual "disk" space used by the content inside the qube itself (500 MiB) is obviously much more than the 8 KiB reported by the qvm- tools.

It seems this happens only with disposables. For AppVMs it is possible to see actual increase of space usage while the qube is running.

**Comment by rustybird** [comment]
>     user@AAA:~ > dd if=/dev/urandom of=~/500mb.img bs=4k iflag=fullblock,count_bytes count=500M

>     user@dom0:~ > qvm-volume info AAA:volatile | grep -E '(size|usage)'

The home directory `~` is on AAA's 'private' volume, not the 'volatile' volume. Even for DisposableVMs.

>     user@dom0:~ qvm-pool info my_ram_pool | grep usage

'private' is a snap_on_start volume for DisposableVMs, so it does not take up space in the pool of AAA but rather in the pool of AAA's disposable template (because the snapshot is created there).

**Comment by brendanhoar** [comment]
> 'private' is a snap_on_start volume for DisposableVMs, so it does not take up space in the pool of AAA but rather in the pool of AAA's disposable template (because the snapshot is created there).

Right. I think this still confuses people who are attempting to use what I'd call "ephemeral pools" for handling sensitive data in "ephemeral qubes". They don't realize the data changes on private/root stay on the source pool.

I think this is somewhat mitigated by recent changes to the home directory being moved to the volatile volume in disposables (con: user no longer controls max volume size where home sits). Or maybe that was whonix only? Haven't checked lately.

Anyway root and private changes still write to whatever pool the templates are in.

As it stands now, you still need to copy a relatively huge template to the ephemeral/disposable pool in order to obtain the type of security wanted for this scenario. This is quite a lot of overhead and isn't a great match for RAM based pools or the lifetimes of multiple independent ephemeral qubes that may run in parallel which should be irrecoverable at vm shutdown.

I tried and failed to devise a method of getting this to work with the 4.x qubes lvm driver quite back, but lvm itself does not support cross device/pool snapshots in a useful way for this.

Raw device-mapper can, I believe, though it's a little tricky. But using that would really require a new qubes file system driver, or some unpleasant device-mapper hackery in the current lvm driver that would likely fail to pass muster.

As it is, my go-to method was to run a script to create a new lvm thin volume in the current pool, stack a crypt device with random key on top, make a new lvm thin pool there, add that to the qubes pools, clone the disposable template (and the normal template it was based on) there (wait a while), make some VM adjustments and then launch one or more the disposable VMs that live fully in this temporary pool. Irrecoverably would have to wait for full tear down of this one pool (or shutdown of the system).

The script could also tear everything down nicely once my project was finished. If I shutdown and forgot, I'd have to force remove all the vm entries, pool reference and the temporary thin volume, only slightly annoying.

B

**Comment by rustybird** [comment]
> I think this is somewhat mitigated by recent changes to the home directory being moved to the volatile volume in disposables (con: user no longer controls max volume size where home sits). Or maybe that was whonix only?

AFAICT no such feature exists? (Tested on R4.2, with a whonix-workstation-17-dvm based DisposableVM)

People might want to focus their energy on generalizing the already existing Qubes OS support for ephemerally encrypted 'root' volumes to 'private' volumes too, which would entail something like the WIP in https://github.com/marmarek/qubes-linux-utils/commit/volatile-thin-pool from, oh wow, 2016. That's a much cleaner approach to anti-forensic DisposableVMs (#904) than creating and tearing down pools.

**Comment by brendanhoar** [comment]
> AFAICT no such feature exists? (Tested on R4.2, with a whonix-workstation-17-dvm based DisposableVM)

Hmm, I inferred it by losing the ability to control the size of the volume that home lived on via the private volume size setting. Obviously something else was going on and I retract the statement.

And yes, what I ended up doing was convoluted but w/o significant more development I didn't see an alternative.

B

**Comment by rustybird** [comment]
> you still need to copy a relatively huge template to the ephemeral/disposable pool

The disposable template itself must be copied (e.g. foo-dvm to foo-copy-dvm), but if you set the ephemeral property of foo-copy-dvm:volatile to True and the rw property of foo-copy-dvm:root to False then [it's okay](https://github.com/QubesOS/qubes-core-admin/commit/24b2891d195ce459b06862c216b1c69d6eef510d) not to copy *the TemplateVM of* foo-dvm.

**Comment by rustybird** [comment]
> According to man qvm-volume:
> 
> "This can be enabled only on a volume with save_on_stop=False and snap_on_start=True - which is only volatile volume."

Good catch, that should be `snap_on_start=False`. If you want to submit a pull request with a fix for this and also


> `qvm-create: error: Pool argument must be of form: -P volume_name=pool_name`

> the correct command seems to be with lowercase '-p'

the relevant files are at

https://github.com/QubesOS/qubes-core-admin-client/blob/main/qubesadmin/tools/qvm_create.py
https://github.com/QubesOS/qubes-core-admin-client/blob/main/doc/manpages/qvm-volume.rst

> Then, when I start the actual DispVM based on that template, I see its volatile volume has ephemeral=False.

Strange, it works for me both on R4.1 and R4.2. Can you post the exact commands you used to set the -dvm properties, launch the DIsposableVM, and check the DisposableVM properties?

**Comment by rustybird** [comment]
> I still don't know how use git, sorry.
> I can open a separate issue for that though. Should I?

No need: https://github.com/QubesOS/qubes-core-admin-client/pull/263

> It is still good to have this ephemeral thing documented better.

Sure. There's a little bit of developer documentation for it at https://dev.qubes-os.org/projects/core-admin/en/latest/qubes-storage.html

Currently the ephemeral property looks more like a building block to eventually develop a user-facing feature with (anti-forensic DisposableVMs), rather than something users are expected to tinker with directly. But even so, it's not documented how ephemeral, rw, the different volume types, the initrd CoW layer, and #904 all fit together.

**Comment by rustybird** [comment]
> https://paste.opensuse.org/pastes/3523df06cbce

Forwarded to https://github.com/QubesOS/qubes-core-admin/pull/564

> So, for all DispVM's the volatile snapshot is always only 8K?

No it depends on how much data has been written to it by the VM. (Swapped memory, and writes that were diverted from a read-only 'root' volume)

> Also, why is the volatile image file always 12G?

There's a per VM class / per volume type *default* size, which is 12 GiB for DispVM / 'volatile':

https://github.com/QubesOS/qubes-core-admin/blob/v4.2.19/qubes/vm/dispvm.py#L80-L81
https://github.com/QubesOS/qubes-core-admin/blob/v4.2.19/qubes/config.py#L60-L61

> And why is that not measured by any conventional size utility (e.g. df)?

`df` and `du` output roughly corresponds to the volumes' usage, not their size.

**Comment by rustybird** [comment]
Yes, `volatile.img` is a [sparse file](https://en.wikipedia.org/wiki/Sparse_file).

**Comment by marmarek** [comment]
You may be interested in `ephemeral_volatile` property on the pool
itself.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by marmarek** [comment]
On Wed, Nov 15, 2023 at 05:04:04AM -0800, emanruse wrote:
> @marmarek
> 
> I have tried that but I am getting:
> 
> qvm-pool: error: Setting option encrypted_volatile is not supported for pool ram_pool_BBB
> 
> It seems it is not possible to use it with tmpfs based pools?

'ephemeral_volatile', not 'encrypted_volatile'

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by marmarek** [comment]
On Wed, Nov 15, 2023 at 05:16:31AM -0800, emanruse wrote:
> > 'ephemeral_volatile'
> 
> There is no such option mentioned anywhere in the man page of qvm-pool.

Right, looks like an error in the man page

> Could you please explain what you mean and how it is used?

It's just a default value for 'ephemeral' value for 'volatile' volumes
created in this pool.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by marmarek** [comment]
On Wed, Nov 15, 2023 at 05:41:31AM -0800, emanruse wrote:
> Thanks for confirming.
> 
> Would you consider ephemeral volume on an ephemeral pool an unnecessary overhead?

Yes, I think so.

> Does it actually add any security (e.g. in case of 2 or more qubes having their volumes on that same pool)?

The ephemeral volatile feature matters when the attacker gets access to
your _unencrypted_ dom0 (either by getting the key, or getting access to
a running unlocked system). In that case tmpfs doesn't guarantee
protection (as it can be written to swap), but ephemeral encryption key
should (while the data may still be there, its encryption key will be long
gone, so it will be just garbage). Using tmpfs to hold those encrypted
volumes doesn't really improve much.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by marmarek** [comment]
> Do I understand correctly that only ephemeral_volatile on the pool itself needs to be used? (and not on the volumes)

Yes.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
